api/server/sdk/credentials.go

```diff
@@ -0,0 +1,404 @@
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/libopenstorage/openstorage/api"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (s *VolumeServer) ProvideForAWS(
+	ctx context.Context,
+	req *api.ProvideCredentialsForAWSRequest,
+) (*api.ProvideCredentialsForAWSResponse, error) {
+
+	if len(req.GetAccessKey()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Access Key")
+	}
+
+	if len(req.GetSecretKey()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Secret Key")
+	}
+
+	if len(req.GetRegion()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Region Key")
+	}
+
+	if len(req.GetEndpoint()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Endpoint Key")
+	}
+
+	params := make(map[string]string)
+
+	params[api.OptCredType] = "s3"
+	params[api.OptCredRegion] = req.GetRegion()
+	params[api.OptCredEndpoint] = req.GetEndpoint()
+	params[api.OptCredAccessKey] = req.GetAccessKey()
+	params[api.OptCredSecretKey] = req.GetSecretKey()
+
+	uuid, err := s.driver.CredsCreate(params)
+
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to create S3 credentials: %v",
+			err.Error())
+	}
+
+	// Validate if the credentials provided were correct or not
+	validateReq := &api.CredentialsValidateRequest{CredentialId: uuid}
+
+	validateErr := s.driver.CredsValidate(validateReq.GetCredentialId())
+
+	if validateErr != nil {
+		deleteCred := &api.CredentialsDeleteRequest{CredentialId: uuid}
+		err = s.driver.CredsDelete(deleteCred.GetCredentialId())
+
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"failed to delete invalid S3 credentials: %v",
+				err.Error())
+		}
+
+		return nil, status.Errorf(
+			codes.Internal,
+			"credentials could not be validated: %v",
+			validateErr.Error())
+
+	}
+	return &api.ProvideCredentialsForAWSResponse{CredentialId: uuid}, nil
+
+}
+
+func (s *VolumeServer) ProvideForAzure(
+	ctx context.Context,
+	req *api.ProvideCredentialsForAzureRequest,
+) (*api.ProvideCredentialsForAzureResponse, error) {
+
+	if len(req.GetAccountKey()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Account Key")
+	}
+
+	if len(req.GetAccountName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply Account name")
+	}
+
+	params := make(map[string]string)
+
+	params[api.OptCredType] = "azure"
+	params[api.OptCredAzureAccountKey] = req.GetAccountKey()
+	params[api.OptCredAzureAccountName] = req.GetAccountName()
+
+	uuid, err := s.driver.CredsCreate(params)
+
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"failed to create Azure credentials: %v",
+			err.Error())
+	}
+
+	// Validate if the credentials provided were correct or not
+	validateReq := &api.CredentialsValidateRequest{CredentialId: uuid}
+
+	validateErr := s.driver.CredsValidate(validateReq.GetCredentialId())
+
+	if validateErr != nil {
```

Comment:
Seems like this delete section is a repetition. Consolidate the processing into one function they can all use.